jetty-util/src/main/java/org/eclipse/jetty/util/thread/Invocable.java

```diff
@@ -33,9 +35,45 @@
 {
     static ThreadLocal<Boolean> __nonBlocking = new ThreadLocal<>();
 
+    /**
+     * <p>The behavior of an {@link Invocable} when it is invoked.</p>
+     * <p>Typically, {@link Runnable}s or {@link Callback}s declare their
+     * invocation type; this information is then used by the code that should
+     * invoke the {@code Runnable} or {@code Callback} to decide whether to
+     * invoke it directly, or submit it to a thread pool to be invoked by
+     * a different thread.</p>
+     */
     enum InvocationType
     {
-        BLOCKING, NON_BLOCKING, EITHER
+        /**
+         * <p>Invoking the {@link Invocable} may block the invoker thread,
+         * and the invocation may be performed immediately (possibly blocking
+         * the invoker thread) or deferred to a later time, for example
+         * by submitting the {@code Invocable} to a thread pool.</p>
+         * <p>This invocation type is suitable for {@code Invocable}s that
+         * call application code, for example to process an HTTP request.</p>
+         */
+        BLOCKING,
+        /**
+         * <p>Invoking the {@link Invocable} does not block the invoker thread,
+         * and the invocation may be performed immediately in the invoker thread.</p>
+         * <p>This invocation type is suitable for {@code Invocable}s that
+         * call implementation code that is guaranteed to never block the
+         * invoker thread.</p>
+         */
+        NON_BLOCKING,
+        /**
+         * <p>Invoking the {@link Invocable} may block the invoker thread,
+         * but the invocation cannot be deferred to a later time, differently
+         * from {@link #BLOCKING}.</p>
+         * <p>This invocation type is suitable for {@code Invocable}s that
+         * themselves produce other {@code Invocable}s.</p>
+         * <p>The invocation of this {@code Invocable} causes the
+         * production of other, nested, {@code Invocable}s, which may be
+         * of type {@link #NON_BLOCKING} and therefore are invoked
+         * immediately, thus advancing a possibly stalled system.</p>
```

Comment:
I think this is too specific.
It is for tasks that can do useful/important work without blocking, but may block in some sub-tasks if allowed to do so by the invocation.